### PR TITLE
Add pattern to fold tensor.empty used by extract in transform dialect

### PR DESCRIPTION
**Pull request by ThomasRaoux**
This allows generating smaller allocations for reduction code generation. Without that we may generate allocation based on the original size rather than the tiled size.

### CODE REVIEW

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I think this already exists upstream.

**Comment by ThomasRaoux** [comment]
> I think this already exists upstream.

ah ok, I looked for it and couldn't find it. Do you have a pointer?

**Comment by ThomasRaoux** [comment]
> > I think this already exists upstream.
> 
> ah ok, I looked for it and couldn't find it. Do you have a pointer?

never mind, I found it. Replaced.

**Review by MaheshRavishankar — APPROVED** [review]
I am stamping, but probably better to wait for Nicolas.

**Comment by nicolasvasilache** [comment]
I asked in offline chat if we really want a blanket application vs more targeted control as it can have non trivial effects on bufferization in general and could be a footgun. @matthias-springer  is the most up-to-date on patterns interference with bufferization and what foldings have "fusion-like" effects.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ struct ApplyPatternsOpPatterns {
   bool eraseUnnecessaryTensorOperands = false;
   bool foldMemrefAliases = false;
   bool foldReassociativeReshapes = false;
+  bool FoldTensorEmpty = false;
```
Comment (matthias-springer) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.h:44]: spelling: lower case

**Review by nicolasvasilache — APPROVED** [review]

**Pull request merged**